sw/device/lib/dif/dif_padctrl.h

```diff
@@ -0,0 +1,147 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_DIF_PADCTRL_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_PADCTRL_H_
+
+#include <stdbool.h>
+
+#include "padctrl_regs.h"
+#include "sw/device/lib/base/mmio.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Pad attributes.
+ *
+ * Some attributes have defined meanings, others are implementation defined
+ * (|kDifPadCtrlAttrAdditionalFirst| and onwards). Not all pads will support all
+ * attributes.
+ */
+typedef enum dif_padctrl_attr {
+  kDifPadctrlAttrIOInvert = 0,
+  kDifPadctrlAttrOpenDrain,
+  kDifPadctrlAttrPullDown,
+  kDifPadctrlAttrPullUp,
+  kDifPadctrlAttrKeeper,
+  kDifPadctrlAttrWeakDrive,
+  kDifPadctrlAttrAdditionalFirst,
+  kDifPadctrlAttrLast = PADCTRL_PARAM_ATTRDW - 1,
+} dif_padctrl_attr_t;
+
+/**
+ * Pad kinds.
+ *
+ * A pad can be an MIO or DIO pad
+ */
+typedef enum dif_padctrl_pad_kind {
+  kDifPadctrlPadMio = 0, /**< MIO (Multiplexed Input/Output) pad. Connected via
+                            pinmux. */
+  kDifPadctrlPadDio,     /**< DIO (Dedicated Input/Output) pad. Connected
+                            directly to peripheral inputs and outputs, bypassing
+                            pinmux. */
+} dif_padctrl_pad_kind_t;
+
+/**
+ * Pad index.
+ *
+ * Combined with |dif_padctrl_pad_kind_t| to refer to a specific pad.
+ */
+typedef uint32_t dif_padctrl_pad_index_t;
+
+/**
+ * Padctrl instance state.
+ *
+ * Padctrl persistent data that is required by the Padctrl API.
+ */
+typedef struct dif_padctrl {
+  mmio_region_t base_addr; /**< Padctrl base address. */
+} dif_padctrl_t;
+
+/**
+ * Padctrl init routine error codes.
+ */
+typedef enum dif_padctrl_init_error {
+  kDifPadctrlInitErrorNone = 0,        /**< Success. */
+  kDifPadctrlInitErrorInvalidPadctrl,  /**< Supplied padctrl_t is not valid. */
+  kDifPadctrlInitErrorRegistersLocked, /**< Peripheral is locked and cannot be
+                                          reset. */
+} dif_padctrl_init_error_t;
+
+/**
+ * Initialises an instance of Padctrl.
+ *
+ * Information that must be retained, and is required to program Padctrl, shall
+ * be stored in @p padctrl.
+ *
+ * @param base_addr Base address of an isntance of the Padctrl IP block
+ * @param padctrl Padctrl state data
+ * @return dif_padctrl_init_error_t.
+ */
+dif_padctrl_init_error_t dif_padctrl_init(mmio_region_t base_addr,
+                                          dif_padctrl_t *padctrl);
+
+/**
+ * Padctrl registers lock routine error codes.
+ */
+typedef enum dif_padctrl_registers_lock_error {
+  kDifPadctrlRegistersLockErrorNone = 0,       /**< Success. */
+  kDifPadctrlRegistersLockErrorInvalidPadctrl, /**< Supplied padctrl_t is not
```

Comment:
I think what you call `None` we've been calling `Ok`, while what you call `InvalidPadctrl` seems to be `BadArg` (being a general "this argument is bad" error).

Also, we've tended to name these types `blah_blah_result` instead of `blah_blah_error`. I can pull up the issue where we discussed this if you'd like.